[SecuritySolution] `Investigate in timeline` from highlighted fields

janmonschke opened the pull request:
## Summary

Re-adds the `Investigate in timeline` button to the highlighted fields table. It now makes sure the timeline metadata is set up correctly. (See https://github.com/elastic/kibana/issues/129591)


https://user-images.githubusercontent.com/68591/165988104-416fda89-d806-471a-8acf-39bb08998e04.mov

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

janmonschke committed:
feat: add "investigate in timeline" btn to prevalence field

The button existed before but it was not creating the timeline correctly as it's described in #129591

janmonschke committed:
fix: make sure the button only shows on hover

janmonschke committed:
chore: remove duplicate `showTimeline` action

janmonschke committed:
test: fix tests

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/investigate-in-timeline-from-prevalence

kibanamachine committed:
Merge branch 'main' into security/investigate-in-timeline-from-prevalence

janmonschke committed:
test: add integration test for investigate in timeline

michaelolo24 commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alerts_details.spec.ts:
```diff
@@ -80,4 +82,12 @@ describe('Alert details with unmapped fields', () => {
       });
   });
+
+  it('Opens a new timeline investigation', () => {
```
Can you test that the number in the flyout matches the number that loads in the timeline?

janmonschke replied:
Sure -> https://github.com/elastic/kibana/pull/131255/commits/983d852a3463597e63dc7e2093740d1c3ca5e9e3

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/table/add_to_timeline_cell.tsx:
```diff
@@ -30,14 +36,48 @@ const AddToTimelineCell = React.memo<AlertSummaryRow['description']>(
+          })
+        );
+        // Lock the time range by default
+        dispatch(inputsActions.removeGlobalLinkTo());
```
Should the lock icon in the timeline display as locked? I think even though the dates matched, it didn't show they were linked

michaelolo24 replied:
Yea that's fine so the user can independently change that range without affecting the global one

janmonschke replied:
Updated: https://github.com/elastic/kibana/pull/131255/commits/f758547524289d6b44dc4efd58f92f232b9d35ce

michaelolo24 replied:
I was just testing it and it looks like it's locked even though the lock icon shows they're not linked

michaelolo24 replied:
Updating the timerange in flyout updates it in the global timerange, but not vice versa

janmonschke replied:
I looked into it and this appears to be a general bug that was not introduced by these changes. I was able to reproduce the same behaviour by starting a new timeline from the alerts table. I would prefer fixing that issue in a separate PR based on a new ticket, wdyt @michaelolo24 ?

michaelolo24 replied:
fyi @kqualters-elastic this might be worth investigating in https://github.com/elastic/kibana/issues/124713

andrew-goldstein replied:
When the new `Investigate in timeline` action for the `Alert prevalence` column is invoked, the status of the lock icon is _unlocked_, which is what I would expect as a user. (This is consistent with the behavior of the row-level  `Investigate in timeline` action in the alerts table.)

janmonschke committed:
test: make sure the event count matches

the event count in the alert flyout and in the timeline should match

janmonschke committed:
chore: update commment

michaelolo24 commented:
Last thing, looks like you'll need to pass the `isReadOnly` prop that's used in `x-pack/plugins/security_solution/public/common/components/event_details/table/summary_value_cell.tsx`  to the add to timeline cell to prevent the button from showing up in the rule preview workflow:

<img width="1305" alt="image" src="https://user-images.githubusercontent.com/17211684/167455458-bb06ce27-78de-4081-821c-3bc303398a5c.png">

janmonschke committed:
fix: hide certain columns in read-only mode

- Do not show the "prevalence" column
- Do not show the "Investigate in timeline" column

janmonschke commented:
@michaelolo24 as discussed offline, both the prevalence column and the investigate in TL column are now hidden in read-only mode: https://github.com/elastic/kibana/pull/131255/commits/8039fe48f352ea83a7a17a9acb72bda48e0561a0

michaelolo24 reviewed (approved):
Thanks for making these changes! 🚀 LGTM!

andrew-goldstein commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alerts_details.spec.ts:
```diff
@@ -80,4 +83,21 @@ describe('Alert details with unmapped fields', () => {
+        cy.get(TIMELINE_TITLE).should('contain.text', 'Untitled timeline');
+
+        // The alert count in this timeline should match the count shown on the alert flyout
+        cy.get(QUERY_TAB_BUTTON).should('contain.text', alertCount);
```
Fields that typically have just one value, like `process.name`, consistently display the same alert counts in both the `Alert prevalence` column in the flyout, and in Timeline.

Fields that have more than one value sometimes display different counts between the flyout and Timeline.

Example: The `process.args` in the sample JSON below contains two values:

```json
    "process": {
      "args": [
        "\"C:\\lsass.exe\"",
        "--jyf"
      ],
      "name": "lsass.exe",
    },
```

In the video below:

- The `process.args` field displays `570` in the `Alert prevalence` column in the flyout
- Timeline displays a count of `285` alerts when the `Investigate in timeline` action is invoked

https://user-images.githubusercontent.com/4459398/167693128-52a0024e-4f71-4308-a8f5-c67050aed8f7.mov

andrew-goldstein replied:
Per a chat with @michaelolo24, we'll track this observation in a separate issue: https://github.com/elastic/kibana/issues/131967

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/table/add_to_timeline_cell.test.tsx:
```diff
@@ -58,7 +58,7 @@ describe('AddToTimelineCellRenderer', () => {
         </TestProviders>
       );
-      expect(screen.getByTestId('test-add-to-timeline')).toBeInTheDocument();
+      expect(screen.queryByLabelText('Investigate in timeline')).toBeInTheDocument();
```
Since `getByTestId` can't be used here, consider adding the same

```ts
import { ACTION_INVESTIGATE_IN_TIMELINE } from '../../../../detections/components/alerts_table/translations';
```

as the component, and updating this (and the other assertion below) to use the imported constant, i.e.

```ts
expect(screen.queryByLabelText(ACTION_INVESTIGATE_IN_TIMELINE)).toBeInTheDocument();
```

andrew-goldstein reviewed (approved):
Thanks for this feature @janmonschke! 🙏  It feels like it may be useful in other parts of the Security Solution in the future.

✅  Desk tested locally

LGTM 🚀

janmonschke committed:
test: use translation string instead of inline string
